driver-core/src/main/com/mongodb/internal/TimeoutContext.java

```diff
@@ -125,8 +131,8 @@ public boolean hasTimeoutMS() {
      * @return true if the timeout has been set and it has expired
      */
     public boolean hasExpired() {
-        // Use timeout.remaining instead of timeout.hasExpired that measures in nanoseconds.
-        return timeout != null && !timeout.isInfinite() && timeout.remaining(MILLISECONDS) <= 0;
+        // TODO-CSOT remove this method (do not inline, but use lambdas)
+        return Timeout.nullAsInfinite(timeout).run(NANOSECONDS, () -> false, (ns) -> false, () -> true);
```

Comment:
This still needs to be done in MS. If there is less than 1MS left then it should not continue, eg shouldnt send anything on the wire or should wrap any seen timeout errors as CSOT errors.